### PR TITLE
Pull Rewrite asm in C into master branch

### PR DESCRIPTION
**Pull request by kevana**

### CODE REVIEW

**Comment by taylortrimble** [comment]
Can we merge master into this so the diff I have to look at is littler? @aterlumen

File: DMXlib.h [hunk 1]
```diff
@@ -0,0 +1,97 @@
+/*
+ * Organization: Tesla Works
+ * Project: Animatronic heads
+ *		Functions for use with PIC18 microcontrollers to receive DMX data
+ *
+ */
+
+// Variables
+char RxBuffer[512];
```
Comment (taylortrimble) [inline DMXlib.h]: Lets do:

``` C
const int DMXBufferSize = 512;
char DMXBuffer[DXMBufferSize];
```

File: DMXlib.h [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+static int numReceivedBytes; //16-bit counter
```
Comment (taylortrimble) [inline DMXlib.h]: `static` in a global variable doesn't make much sense! Lets use `int DMXBytesReceived`.

File: DMXlib.h [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+char inputBuffer; //used to read RCREG to clear error conditions
```
Comment (taylortrimble) [inline DMXlib.h]: `char DMXInputBuffer;`

File: DMXlib.h [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+const int DMXStartCode = 0x00;
+enum {
+    MainLoop, WaitBreak, GotBreak, WaitForStart, WaitForData, MoveData, Done
```
Comment (taylortrimble) [inline DMXlib.h]: Can we do:

``` C
enum {
    DMXMainLoop, DMXWaitBreak, DMXGotBreak, DMXWaitForStart, DMXWaitForData, DMXMoveData, DMXDone
} DMXState;
```

:question: 
It's a little more verbose, but more framework-y styled.

File: DMXlib.h [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+} state;
+
+void setupDMX(void)
```
Comment (taylortrimble) [inline DMXlib.h]: `void DMXSetup(void)`

File: DMXlib.h [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+{
+    OSCTUNEbits.PLLEN = 1; //Set PLL on. Not sure why this isn't in the main setup function
+    for (int i = 0; i < 512; i++) { //Clear the receive buffer
```
Comment (taylortrimble) [inline DMXlib.h]: `512` :arrow_right: `DMXBufferSize`

**Comment by taylortrimble** [comment]
Holy :poop: 5ae84f696249b356cfafcaa1d06f362bfcd6d76c is one of the most beautiful commits I've ever seen. That's some **_sexy_** code!

I've added a few initial line comments on style, along the lines of:

```
const int DMXConstantName;
int DMXGlobalName;
void DMXFunctionName(void) {}
```

To make it more framework-like. I'll go over it for accuracy later. But holy shit. Nice work @aterlumen!

**Comment by taylortrimble** [comment]
Also, maybe we could move the code out of the header file? :stuck_out_tongue_winking_eye:

**Comment by kevana** [comment]
I think unless @tylrtrmbl finds any more logic errors in DMXlib.h this is ready to pull into master.

File: DMXlib.c [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+const int DMXStartCode = 0x00;
+enum {
+    DMXMainLoop, DMXWaitBreak, DMXGotBreak, DMXWaitForStart, DMXWaitForData, DMXDone
+} DMXState;
+
+void DMXSetup(void)
+{
+    OSCTUNEbits.PLLEN = 1; //Set PLL on. Not sure why this isn't in the main setup function
+    for (int i = 0; i < DMXBufferSize; i++) { //Clear the receive buffer 
+    	DMXBuffer[i] = 0;
```
Comment (taylortrimble) [inline DMXlib.c]: Alignment; also on line 26, 27, 30

File: DMXlib.c [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+ *
+ * Author: Kevan Ahlquist (@aterlumen)
+ */
+
+// Variables
+const int DMXBufferSize = 512;
+char DMXBuffer[DMXBufferSize];
+int DMXBytesReceived; //16-bit counter
+char DMXInputBuffer; //used to read RCREG to clear error conditions
+const int DMXStartCode = 0x00;
```
Comment (taylortrimble) [inline DMXlib.c]: Can we have separate

```
// Constants
...

// Variables
...
```

:question:

File: DMXlib.c [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+enum {
+    DMXMainLoop, DMXWaitBreak, DMXGotBreak, DMXWaitForStart, DMXWaitForData, DMXDone
+} DMXState;
+
+void DMXSetup(void)
+{
+    OSCTUNEbits.PLLEN = 1; //Set PLL on. Not sure why this isn't in the main setup function
```
Comment (taylortrimble) [inline DMXlib.c]: Let's move this there.

File: DMXlib.h [hunk 10]
```diff
@@ -0,0 +1,16 @@
+/*
+ * Organization: Tesla Works
+ * Project: Animatronic heads
+ *		Functions for use with PIC18 microcontrollers to receive DMX data
+ *
+ * Author: Kevan Ahlquist (@aterlumen)
+ */
+ 
+ // Variables
+const int DMXBufferSize = 512;
```
Comment (taylortrimble) [inline DMXlib.h]: We actually don't want this or line 11 declared twice; we should remove them from the `.c` file.

File: DMXlib.c [hunk 11]
```diff
@@ -0,0 +1,101 @@
+/*
+ * Organization: Tesla Works
+ * Project: Animatronic heads
+ *		Functions for use with PIC18 microcontrollers to receive DMX data
+ *
+ * Author: Kevan Ahlquist (@aterlumen)
+ */
+
```
Comment (taylortrimble) [inline DMXlib.c:8]: `DMXlib.c` should `#include "DMXlib.h"`.

**Comment by taylortrimble** [comment]
This is one of the best code ports I've ever seen. Hurrah! after all the stylistic, spacing, and other fixes I mentioned in the last diff and 51d5916bf1357f2d695b44270cfc423e765bc8fd are done, it's good to pull!
:shipit:

**Comment by kevana** [comment]
@tylrtrmbl One last look? I think we're good to go here.

**Comment by taylortrimble** [comment]
:sparkles:

**Pull request merged**